hw/opentitan/ot_csrng.c

```diff
@@ -666,59 +762,67 @@ static void ot_csrng_drng_update(OtCSRNGInstance *inst)
 
+    unsigned slot = ot_csrng_get_slot(inst);
     drng->seeded = false;
 
     if (!flag0) {
-        if (!inst->parent->es_available) {
+        if (!s->es_available) {
             qemu_log_mask(LOG_GUEST_ERROR,
-                          "%s: Cannot request entropy w/o power cycling ES\n",
+                          "%s: Requesting entropy w/o power cycling ES\n",
                           __func__);
-            return CSRNG_CMD_ERROR;
+            /*
+             * Continue anyway as it seems HW does not enforce what is
+             * documented. Force the flag so the warning message is only
+             * shown once (it does not serve any other purpose).
+             */
+            s->es_available = true;
         }
 
         uint64_t buffer[OT_RANDOM_SRC_DWORD_COUNT];
         memset(buffer, 0, sizeof(buffer));
         unsigned len = drng->material_len * sizeof(uint32_t);
         memcpy(buffer, drng->material, MIN(len, sizeof(buffer)));
 
-        OtCSRNGState *s = inst->parent;
+
         uint64_t entropy[OT_RANDOM_SRC_DWORD_COUNT];
         int res;
         bool fips;
-        trace_ot_csrng_request_entropy(ot_csrng_get_slot(inst),
-                                       s->entropy_gennum);
+        trace_ot_csrng_request_entropy(slot, s->entropy_gennum);
         OtRandomSrcIfClass *cls = OT_RANDOM_SRC_IF_GET_CLASS(rand_dev);
         OtRandomSrcIf *randif = OT_RANDOM_SRC_IF(rand_dev);
         res = cls->get_random_values(randif, s->entropy_gennum, entropy, &fips);
         if (res) {
             s->entropy_delay = (res > 1) ? (unsigned)res : 0;
-            trace_ot_csrng_entropy_rejected(ot_csrng_get_slot(inst),
+            trace_ot_csrng_entropy_rejected(slot,
                                             res < 0 ? (res == -2 ? "stalled" :
                                                                    "error") :
                                                       "not ready",
                                             res);
-            return res < 0 ? (res == -2 ? CSRNG_CMD_STALLED : CSRNG_CMD_ERROR) :
+            return res < 0 ? (res == -2 ? CSRNG_CMD_STALLED :
+                                          CSRNG_CMD_RESEED_CNT_EXCEEDED) :
                              CSRNG_CMD_RETRY;
         }
+
         /* always perform XOR which is a no-op if material_len is zero */
         for (unsigned ix = 0; ix < OT_RANDOM_SRC_DWORD_COUNT; ix++) {
             buffer[ix] ^= entropy[ix];
         }
         memcpy(drng->material, buffer, sizeof(entropy));
         drng->material_len = sizeof(entropy) / (sizeof(uint32_t));
         ot_csrng_drng_update(inst);
-        drng->fips = fips;
+        drng->no_fips |= !fips;
     } else {
         ot_csrng_drng_update(inst);
-        drng->fips = false;
+        drng->no_fips = true;
     }
 
-    drng->reseed_counter = 1u;
-
+    drng->reseed_counter = 0u;
     drng->seeded = true;
+    drng->force_fips = (bool)(s->regs[R_FIPS_FORCE] >> slot);
```

Comment:
Maybe I'm misunderstanding, but shouldn't this be something like
```suggestion
    drng->force_fips = (bool)((s->regs[R_FIPS_FORCE] >> slot) & 0x1);
```
as otherwise for example forcing the FIPS of just instance 2 would also lead this to be true for slot 1, despite the fact that instance 1 itself might be disabled? I'm not too familiar with the CSRNG so I may just not have a full understanding here.

Reply:
Oops. good catch